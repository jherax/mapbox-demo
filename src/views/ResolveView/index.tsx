import {type ReactNode} from 'react';
import styled from 'styled-components';

const Wrapper = styled.section`
  padding: 50px;
`;

export default function ResolveView(props: ResolveViewProps) {
  if (props.loading) {
    return (
      <Wrapper className='ResolveView_loading'>
        <pre>Loading...</pre>
      </Wrapper>
    );
  }

  if (props.error) {
    if (typeof props.beforeRenderError === 'function') {
      props.beforeRenderError();
    }
    return (
      <Wrapper className='ResolveView_error'>
        <pre>{props.error.message}</pre>
      </Wrapper>
    );
  }

  return <>{props.children}</>;
}

export type ResolveViewProps = Readonly<
  Partial<{
    loading: boolean;
    error: {
      message: string;
    };
    children: ReactNode;
    beforeRenderError: () => void;
  }>
>;
