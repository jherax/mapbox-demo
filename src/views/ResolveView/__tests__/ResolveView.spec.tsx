import {render, screen} from '@testing-library/react';

import ResolveView from '../index';

describe('Testing <ResolveView />', () => {
  it('should render the default view correctly', async () => {
    const {asFragment} = render(
      <ResolveView>
        <p>Default View</p>
      </ResolveView>,
    );

    expect(await screen.findByText('Default View')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render correctly when the loading is true', async () => {
    const {asFragment} = render(
      <ResolveView loading={true}>
        <p>Default View</p>
      </ResolveView>,
    );

    expect(await screen.findByText('Loading...')).toBeInTheDocument();
    expect(screen.queryByText('Default View')).not.toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render correctly when an error is present', async () => {
    const error = new Error('Something went wrong');
    const beforeError = jest.fn();
    const {asFragment} = render(
      <ResolveView error={error} beforeRenderError={beforeError}>
        <p>Default View</p>
      </ResolveView>,
    );

    expect(await screen.findByText('Something went wrong')).toBeInTheDocument();
    expect(screen.queryByText('Default View')).not.toBeInTheDocument();
    expect(beforeError).toHaveBeenCalledTimes(1);
    expect(asFragment()).toMatchSnapshot();
  });
});
