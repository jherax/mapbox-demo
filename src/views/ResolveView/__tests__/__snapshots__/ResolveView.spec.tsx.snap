// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <ResolveView /> should render correctly when an error is present 1`] = `
<DocumentFragment>
  <section
    class="sc-bdvuGq foGWrU"
  >
    <pre>
      Something went wrong
    </pre>
  </section>
</DocumentFragment>
`;

exports[`Testing <ResolveView /> should render correctly when the loading is true 1`] = `
<DocumentFragment>
  <section
    class="sc-bdvuGq foGWrU"
  >
    <pre>
      Loading...
    </pre>
  </section>
</DocumentFragment>
`;

exports[`Testing <ResolveView /> should render the default view correctly 1`] = `
<DocumentFragment>
  <p>
    Default View
  </p>
</DocumentFragment>
`;
